[RFC] Introduce a RateLimiter component

wouterj opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Refs #37444 
| License       | MIT
| Doc PR        | tbd

Based on the discussions in #37444, I decided to write a general purpose RateLimiter component. This implementation uses the token bucket algorithm, inspired by the [Go's time/rate](https://github.com/golang/time/blob/3af7569d3a1e776fc2a3c1cec133b43105ea9c2e/rate/rate.go) library and the [PHP `bandwidth-throttle/token-bucket` package](https://github.com/bandwidth-throttle/token-bucket) (which is [unmaintained for years](https://github.com/bandwidth-throttle/token-bucket/issues/19)).

### Usage

The component has two main methods:

* `Limiter::reserve(int $tokens, int $maxTime)`, allocates `$tokens` and returns a `Reservation` containing the wait time. Use this method if your process wants to wait before consuming the token.
* `Limiter::consume(int $tokens)`, checks if `$tokens` are available now and discards the reservation if that's not the case. Use this method if you want to skip when there are not enough tokens at this moment.

The component uses the Lock component to make sure it can be used in parallel processes.

Example:

```php
<?php

namespace App\EventListener;

use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\Event\RequestEvent;
use Symfony\Component\RateLimiter\LimiterFactory;

class LimitListener
{
    private $limiterFactory;

    public function __construct(LimiterFactory $apiLimiterFactory)
    {
        $this->limiterFactory = $apiLimiterFactory;
    }

    public function __invoke(RequestEvent $event)
    {
        $ip = $event->getRequest()->getClientIp();

        $limiter = $this->limiterFactory->createLimiter(preg_replace('/[^a-zA-Z0-9]/', '-', $ip));
        if (!$limiter->consume()) {
            $event->setResponse(new Response('Too many requests', 429));
        }
    }
}
```

### Usefullness of the component

I think a generic rate limiter is usefull in quite some places:

* Add a login throttling feature in Symfony
* <s>Rate limiting outgoing API calls (e.g. HttpClient), to prevent hitting upstream API limits.</s> See #37471 (and https://blog.heroku.com/rate-throttle-api-client )
* Allowing users to easily implement API rate limits in their own Symfony-based APIs.

### State of the art

There are some rate limiting packages in PHP, but I think there is no precendent for this component:

* [`graham-campbell/throttle`](https://github.com/GrahamCampbell/Laravel-Throttle) is heavily relying on Laravel. It is however very popular, proofing there is a need for such feature
* [`nikolaposa/rate-limit`](https://github.com/nikolaposa/rate-limit) does not implement reservation of tokens and as such less feature complete. Also its architecture combines the rate limiter and storage, making it harder to implement different storages.

### Todo

If it is agreed that this component can bring something to Symfony, it needs some more love:

* [x] Add more tests
* [x] Integrate with the FrameworkBundle
* [x] Add sliding window implementation
* [x] Add integration with the Security component
* <s>Maybe add more storage implementations? I didn't want to duplicate storage functionalities already existing in the Lock and Cache component, thus I for now focused mostly on integrating the Cache adapters. But maybe a special Doctrine adapter makes sense?</s>

wouterj committed:
[RFC] Introduce a RateLimiter component

fabpot commented:
Thanks for working on this. I agree that having such a component in core would be wonderful.

fabpot commented on src/Symfony/Component/RateLimiter/LimiterFactory.php:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class LimiterFactory
```
I'm not sure we need such a class.

wouterj replied:
Hmm, I think we need such a class for integration in Symfony. I would love to allow users to configure limiters in the config, e.g:

```yaml
framework:
  limiter:
    api:
      storage: ...
      rate: { hour: 10 }
      initial: 10
```

Based on this limiter, a specific limiter with an id should be created (e.g. client IP for limiter ids). I.e:

```php
public function someApiAction(LimiterFactory $factory, Request $request)
{
    $limiter = $factory->createLimiter('api', $request->getClientId());
    if (!$limiter->consume()) {
        return new JsonResponse(0, 429);
    }

   // ...
}
```

fabpot commented on src/Symfony/Component/RateLimiter/TokenBucket.php:
```diff
@@ -0,0 +1,105 @@
+ *
+ * @internal
+ */
+class TokenBucket
```
final?

fabpot commented on src/Symfony/Component/RateLimiter/TokenBucket.php:
```diff
@@ -0,0 +1,105 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @internal
```
Why is it internal?

wouterj replied:
If we're going to keep the factory, the token bucket is only used in `LimiterFactory` and `Limiter` - not in any userland code.

fabpot commented on src/Symfony/Component/RateLimiter/Rate.php:
```diff
@@ -0,0 +1,95 @@
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
```
I'd prefer to use `final class Rate` instead. Any reasons not to do so?

wouterj replied:
I thought Symfony never used PHP's final. Fixed

fabpot commented on src/Symfony/Component/RateLimiter/Rate.php:
```diff
@@ -0,0 +1,95 @@
+
+    public static function perHour(int $rate): self
+    {
+        return new static($rate / 3600, self::UNIT_HOUR);
```
When I want a Rate of 3600 per hour, it's not the same as 1 per second. In the first case, consuming 3600 tokens right away should be possible, and then you cannot consume anything anymore for about an hour minus 1 second.

I might have missed something, but it looks like here you are enforcing 1 per second.

wouterj replied:
Agreed, this is a missing feature. The Go package I used as a reference has this limitation. We should keep the original unit and split "fill time" and "fill amount" as described in https://medium.com/smyte/rate-limiter-df3408325846

fabpot commented on src/Symfony/Component/RateLimiter/Storage/StorageInterface.php:
```diff
@@ -0,0 +1,23 @@
+interface StorageInterface
+{
+    public function save(TokenBucket $bucket): void;
+    public function fetch(string $bucketId): ?TokenBucket;
```
I would not add a Doctrine adapter at first. Using the Cache component seems enough to me.

noniagriconomie replied:
@fabpot ok i've seen the `Symfony\Component\Cache\Adapter\DoctrineAdapter` so it should be good i think

stof commented:
>     * Rate limiting outgoing API calls (e.g. HttpClient), to prevent hitting upstream API limits.

Based on the naming of https://schneems.com/2020/06/25/rate-limiting-rate-throttling-and-how-they-work-together/, what HttpClient would need is a rate throttling implementation, not a rate limiting one (which does not exactly work the same).
Heroku published a blog post recently about their rate throttling algorithm in their SDK btw: https://blog.heroku.com/rate-throttle-api-client

Are you sure that the same component can be used for both ?

wouterj commented on src/Symfony/Component/RateLimiter/Limiter.php:
```diff
@@ -0,0 +1,105 @@
+     */
+    public function reserve(int $tokens = 1, ?float $maxTime = null): Reservation
+    {
+        if ($tokens > $this->maxBurst) {
```
`$maxBurst` should be an int, fixed

wouterj commented:
> Are you sure that the same component can be used for both ?

Yes, I think so, that's where the two different methods come into play. The example shown in the PR description is a throttling example (it throttles the number of time echo's we are creating).

An example of a rate limiter can be:

```php
<?php

require 'vendor/autoload.php';

$request = new \Symfony\Component\HttpFoundation\Request::createFromGlobals();

$factory = new \Symfony\Component\RateLimiter\LimiterFactory();
$limiter = $factory->createLimiter(10, \Symfony\Component\RateLimiter\Rate::perSecond(1), $request->getClientIp());

if (!$limiter->consume()) {
    (new \Symfony\Component\HttpFoundation\Response(null, 429))->send();
}

// ...
```

stof commented:
@wouterj but for an HttpClient, the proper implementation of rate throttling is to adapt to the rate limits of the target API, not to implement rate limits on its own (as it might not be the only client involved to consume the rate limits) That's why I'm doubting that the same component would be used for that.

wouterj commented:
Ah, I've now read the Heroku article you posted and I see what you mean. I still think this component works for both throttling and limiting, but API call throttling might need a different approach indeed. For the focus of this PR, let's keep the API throttling discussions in #37471.

derrabus commented on src/Symfony/Component/RateLimiter/NoLock.php:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @final
+ */
+class NoLock implements LockInterface
```
Shouldn't this class be part of the Lock component?

wouterj replied:
Yes, probably.

kbond commented:
Heh, interesting timing! I'm just finalizing a *generic* throttling library of my own: https://github.com/zenstruck/governator

It is basically a port of the throttling system from Laravel (core) and uses the *fixed window* strategy. I used the lock component for inspiration on configuring the *stores*.

This certainly looks more robust and would love to see rate limiting in Symfony core!

wouterj commented:
This PR now includes integration into the FrameworkBundle using a `LimiterFactory` (this copies the approach of the Lock component), see updated example in the PR description.

I've come to the conclusion that a generic limiter component must at least provide 2 different limiter algorithms. I propose the following 2 algorithms:

* **Token bucket** (originally in this PR), allows limiting bursts and rates: "After 5 failed attempts, you can only try once every 15 minutes."
* **Sliding window**, common in API limiting: "You can make a maximum of 50 requests per hour"

Other possible algorithms are: leaky bucket (more like a queue) or fixed window (implemented by Laravel). See https://medium.com/figma-design/an-alternative-approach-to-rate-limiting-f8a06cf7c94c and https://blog.cloudflare.com/counting-things-a-lot-of-different-things/ on reasons to use sliding window instead of the fixed window implemented in Laravel.

wouterj commented:
So I've been coming back to this PR locally a couple of times and I'm a bit stuck.

The component as-is provides a working Token Bucket limiter implementation. This works great for something like login throttling, but might not be the best limiter algorithm for API limits (sliding window/fixed window is a better choice). However, I can't find a way to add the same "reservation" support to these window algorithms. Where I need help: **Should we provide one limiter interface for different algorithms, or should we create 2 separate interfaces?**

Also, using the cache component for storage is nice, but directly implementing e.g. Redis can improve performance significantly I think (as you can use REdis commands directly, instead of a "fetching -> php code -> storing" loop). I think that's out of the initial scope of this PR, but something to be taken into account.

fabpot commented:
I don't see any problem in providing 2 interfaces.
Regarding Redis, we can probably improve performance in another PR.

azjezz commented on src/Symfony/Component/RateLimiter/CompoundLimiter.php:
```diff
@@ -0,0 +1,38 @@
+    public function consume(int $tokens = 1): bool
+    {
+        $allow = true;
+        foreach ($this->limiters as $i => $limiter) {
```
```suggestion
        foreach ($this->limiters as $limiter) {
```

azjezz commented on src/Symfony/Component/RateLimiter/CompoundLimiter.php:
```diff
@@ -0,0 +1,38 @@
+    {
+        $allow = true;
+        foreach ($this->limiters as $i => $limiter) {
+            $allow = $limiter->consume($tokens) && $allow;
```
should this be: 

```suggestion
            $allow = $allow && $limiter->consume($tokens);
```

? ( short circuit )

wouterj replied:
This changes the behavior of the compound limiter (as `consume()` is not called when `$allow` is already false). I'm not sure about the expected behavior with compound limiters, but counting all hits on all limiters is at least consistent (otherwise, some limiters will be called and others won't, unless we sort them with the most likely to reach its it limit as first).

dunglas commented:
Regarding web API limits, isn't it better to handle them at the edge of the network? Most web servers (NGINX, Apache) and proxies such as Cloudflare support fixing rate limits. Enforcing the limits at this level prevent to start a PHP process for nothing.

wouterj commented:
This PR is now ready in my opinion. I've moved the security integration using login throttling to another PR: https://github.com/symfony/symfony/pull/38204 (to not block this PR too much)

I've also implement a simpler to implement variable Fixed Window strategy, rather than the Sliding Window. Laravel also only has fixed window strategies. This means one could have a burst of hits at the end of one window, and at the start of the next. As shown in this image (source: https://medium.com/figma-design/an-alternative-approach-to-rate-limiting-f8a06cf7c94c):

![image](https://miro.medium.com/max/749/1*OtN-jP5mZko-OUsKzN28Gg.png)

However, I've also introduced a Compound limiter. Using multiple fixed window limiters, you can somewhat flatten out the bursts as well.

(Fabbot failure appears to be unrelated)

wouterj commented:
> Regarding web API limits, isn't it better to handle them at the edge of the network? Most web servers (NGINX, Apache) and proxies such as Cloudflare support fixing rate limits. Enforcing the limits at this level prevent to start a PHP process for nothing.

Yes, I agree - that is for larger apps. In the documentation of this component, we have to be carefull to place this component in the world of rate limiting on other elements of the network. E.g. I can imagine quite some apps build with Symfony that aren't run behind Cloudflare or on a cloud hosting. Those might just use this for API limits, as long as they are aware of the disadvantage (still having to boot the Symfony kernel).

fabpot commented on src/Symfony/Component/Lock/NoLock.php:
```diff
@@ -0,0 +1,53 @@
+/**
+ * A non locking lock.
+ *
+ * This is used when no mutex is required for this rate limiter
```
This comment should probably be removed or modified as the class is not in the RateLimiter component anymore.

fabpot commented on src/Symfony/Component/Lock/NoLock.php:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
```
Should be a "real" final class instead.

fabpot commented on src/Symfony/Component/RateLimiter/LimiterFactory.php:
```diff
@@ -0,0 +1,97 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
```
final class instead

fabpot commented on src/Symfony/Component/RateLimiter/LimiterFactory.php:
```diff
@@ -0,0 +1,97 @@
+ *
+ * @final
+ */
+class LimiterFactory
```
Can we rename it to `Limiter`?

wouterj replied:
Hmm, I'm not completely sure about this. At the moment, I've used the same tactics as the Lock component (which also has a factory + limiters) and register 2 autowirings for each limiter:

* `LimiterInterface $<limiterName>Limiter`, which injects `$limiterFactory->createLimiter()` (i.e. an application wide limiter)
* `LimiterFactory $<limiterName>LimiterFactory`, which injects `$limiterFactory`, so a class can define a specific limiter key (see the [login throttling implementation](https://github.com/symfony/symfony/pull/38204/files#diff-e0d15bf4273820e4a8f71be498797fc7R49-R51))

If we call this one `Limiter`, you will get these autowiring config:

* `LimiterInterface $<limiterName>Limiter`
* `Limiter $<limiterName>Limiter`

Or should I instead also drop the application-wide limiter autowiring (so a user would always need to do `$limiter->create()` or `$limiter->create(...)`)?

fabpot replied:
I'm not sure that the application-wide limiter autowiring is that useful, but I might be wrong.

stof replied:
the other issue we have is that we have a `Limiter` class which is then not an implementation of the `LimiterInterface`, which can be quite confusing.

fabpot commented on src/Symfony/Component/RateLimiter/LimiterFactory.php:
```diff
@@ -0,0 +1,97 @@
+        $this->config = $options->resolve($config);
+    }
+
+    public function createLimiter(?string $key = null): LimiterInterface
```
might be `create()` instead (`Limiter::create()` reads better than `LimiterFactory::createLimiter()`)

fabpot commented on src/Symfony/Component/RateLimiter/LimiterFactory.php:
```diff
@@ -0,0 +1,97 @@
+                return new FixedWindowLimiter($id, $this->config['limit'], $this->config['interval'], $this->storage, $lock);
+
+            default:
+                throw new \LogicException(sprintf('Limiter strategy "%s" does not exists, it must be either "token_bucket" or "fixed_window".', $this->config['strategy']));
```
typo: exist (no s)

fabpot commented on src/Symfony/Component/RateLimiter/LimiterFactory.php:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    protected static function configureOptions(OptionsResolver $options): void
```
As the class is final, should be private I suppose.

fabpot commented on src/Symfony/Component/RateLimiter/Reservation.php:
```diff
@@ -0,0 +1,45 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
```
final class

fabpot commented on src/Symfony/Component/RateLimiter/TokenBucket.php:
```diff
@@ -0,0 +1,84 @@
+    private $timer;
+
+    /**
+     * TokenBucket constructor.
```
Can be removed

fabpot commented on src/Symfony/Component/RateLimiter/TokenBucketLimiter.php:
```diff
@@ -0,0 +1,115 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class TokenBucketLimiter implements LimiterInterface
```
Could be marked as final?

fabpot commented on src/Symfony/Component/RateLimiter/TokenBucketLimiter.php:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    /**
+     * Use this method if you intend to wait until the required number
```
Waits until the required number of tokens is available.

fabpot commented on src/Symfony/Component/RateLimiter/Window.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class Window implements LimiterStateInterface
```
final?

fabpot commented on src/Symfony/Component/RateLimiter/CompoundLimiter.php:
```diff
@@ -0,0 +1,45 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class CompoundLimiter implements LimiterInterface
```
final class?

fabpot commented on src/Symfony/Component/RateLimiter/FixedWindowLimiter.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class FixedWindowLimiter implements LimiterInterface
```
final class? All classes should be final in this component maybe?

fabpot commented on src/Symfony/Component/RateLimiter/README.md:
```diff
@@ -0,0 +1,41 @@
+
+The Rate Limiter component provides a Token Bucket implementation to
+rate limit input and output in your application.
+
```
You should mark the component as experimental:

```
**This Component is experimental**.
[Experimental features](https://symfony.com/doc/current/contributing/code/experimental.html)
are not covered by Symfony's
[Backward Compatibility Promise](https://symfony.com/doc/current/contributing/code/bc.html).
```

fabpot reviewed (changes requested):
The component should be marked as experimental (via `@experimental in 5.2` in all classes).

wouterj commented:
Alright, I've updated the PR conform your comments (I think I agree that application wide limiters are probably quite rare).

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1707,4 +1709,41 @@ private function addNotifierSection(ArrayNodeDefinition $rootNode)
+                        ->then(function (array $v) { return ['limiters' => $v]; })
+                    ->end()
+                    ->children()
+                        ->arrayNode('limiters')
```
```
framework:
    rate_limiter:
        limiters:
            # allow up to 5 failed login attempts, before allowing 1 every 15 minutes
            login:
```

The `rate_limiter` and THEN `limiters` looks unfortunate. Is it possible to flatten it?

```
framework:
    rate_limiters:
        # allow up to 5 failed login attempts, before allowing 1 every 15 minutes
        login:
```

I can't think of any "system-wide rate limiter" config that might be needed in the future... but tell me if you disagree.

wouterj replied:
Oh, in fact this config does allow skipping the "limiters". So:

```yaml
framework:
    rate_limiter:
        login:
            # ...
```

It can be usefull for forward-compatibility and XML-support:

```xml
<framework:config>
    <framework:rate-limiter>
        <framework:limiter name="..." strategy="..." limit="5"/>
        <framework:limiter name="..." strategy="..." limit="5"/>
    </framework:rate-limiter>
</framework:config>
```

stof replied:
having a sub-node for the prototyped array is useful only for forward-compat to allow having other config settings in `rate_limiter`. XML compat is not an issue. Having `rate_limiters` in YAML would translation to multiple `<framework:rate-limiter>` nodes

fabpot replied:
Everything is fine like this then, right?

wouterj replied:
Yes, I think so.

@stof yes, but imho it's nicer to have one "main" element (`<rate-limiter>`) with multiple sub elements :)

fabpot commented:
Thank you @wouterj.

kdefives commented:
I know this PR is already merge but do you know if it is a good idea to use this feature to build a RateLimiter middleware for Guzzle Http client? 
I explain how I use this experimental feature here: https://codereview.stackexchange.com/q/257312/239259

Any feedback appreciated :)
